File: bot.js
const { Telegraf } = require('telegraf');
const { google } = require('googleapis');
const sheets = google.sheets('v4');

// Environment variables
const botToken = process.env.BOT_TOKEN;
const spreadsheetId = process.env.SHEET_ID;
const serviceAccountEmail = process.env.SERVICE_ACCOUNT_EMAIL;
const privateKey = process.env.PRIVATE_KEY.replace(/\\n/g, '\n');

// Google Sheets auth
const auth = new google.auth.JWT(
  serviceAccountEmail,
  null,
  privateKey,
  ['https://www.googleapis.com/auth/spreadsheets']
);

// Initialize Telegram bot
const bot = new Telegraf(botToken);

// Function to get membership data from Google Sheets
async function getMembershipData(cardNumber) {
  const request = {
    spreadsheetId,
    range: 'Sheet1!A:C',  // Adjust the range to where your data is
    auth,
  };
  
  const response = await sheets.spreadsheets.values.get(request);
  const rows = response.data.values;
  
  if (rows.length) {
    for (let row of rows) {
      if (row[0] === cardNumber) {  // Assuming card number is in the first column
        return { cardNumber: row[0], tier: row[1] }; // Tier is in the second column
      }
    }
  }
  return null;
}

// Bot command: /start
bot.start((ctx) => ctx.reply('Welcome! Please enter your membership card number.'));

// Handle membership validation
bot.on('text', async (ctx) => {
  const cardNumber = ctx.message.text;
  
  const member = await getMembershipData(cardNumber);
  
  if (member) {
    ctx.reply(`Welcome ${member.tier} member!`);
    // Add logic to navigate the user based on their membership tier
  } else {
    ctx.reply('Invalid membership card number.');
  }
});

bot.launch();

File: bot.py
import os
import logging
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from telegram import Update
from telegram.ext import ApplicationBuilder, CommandHandler, MessageHandler, filters, ContextTypes

# Set up logging
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)

# Google Sheets API configuration
SCOPES = ["https://www.googleapis.com/auth/spreadsheets.readonly"]
SAMPLE_SPREADSHEET_ID = "1m0wrn2aaCY8wdKIFpxKg_k1O8ZMf1impeOIfAymcMnw"  # Replace with your spreadsheet ID
SAMPLE_RANGE_NAME = "Members!A2:C"  # Adjust range as necessary

# Initialize credentials
creds = None
if os.path.exists("token.json"):
    creds = Credentials.from_authorized_user_file("token.json", SCOPES)
if not creds or not creds.valid:
    if creds and creds.expired and creds.refresh_token:
        creds.refresh(Request())
    else:
        flow = InstalledAppFlow.from_client_secrets_file("credentials.json", SCOPES)
        creds = flow.run_local_server(port=0)
    with open("token.json", "w") as token:
        token.write(creds.to_json())

# Function to authenticate user
def authenticate_member(card_number):
    try:
        service = build("sheets", "v4", credentials=creds)
        sheet = service.spreadsheets()
        result = sheet.values().get(spreadsheetId=SAMPLE_SPREADSHEET_ID, range=SAMPLE_RANGE_NAME).execute()
        values = result.get("values", [])

        for row in values:
            if row[0] == card_number:  # Assuming card number is in the first column
                return row[1]  # Return membership tier or any other info needed
        return None
    except HttpError as err:
        logging.error(err)
        return None

# Command handler to start the bot
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text("Welcome! Please enter your membership card number.")

# Message handler to process card number
async def handle_message(update: Update, context: ContextTypes.DEFAULT_TYPE):
    card_number = update.message.text
    membership_tier = authenticate_member(card_number)

    if membership_tier:
        await update.message.reply_text(f"Authenticated! Your membership tier is: {membership_tier}")
        # Here you can add logic to navigate based on membership tier
    else:
        await update.message.reply_text("Authentication failed. Please check your card number.")

# Main function to run the bot
def main():
    application = ApplicationBuilder().token("YOUR_TELEGRAM_BOT_TOKEN").build()  # Replace with your bot token

    application.add_handler(CommandHandler("start", start))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_message))

    application.run_polling()

if __name__ == "__main__":
    main()
